### PR TITLE
HPCC-25098 Spot invalid single colon logical filenames

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@mckellyln - please review

File: dali/base/dautils.cpp [hunk 1]
```diff
@@ -417,7 +417,20 @@ void CDfsLogicalFileName::normalizeName(const char *name, StringAttr &res, bool
...
         {
             case '@': ct = s; break;
-            case ':': ct = nullptr; break;
+            case ':':
+            {
+                if (*(s+1) == ':')
+                {
+                    ++s;
+                    char nextc=*(s+1);
+                    if (!validFNameChar(nextc) && (!allowWild || ('*' != nextc) && ('?' != nextc) && ('{' != nextc)))
```
Comment (mckellyln) [inline dali/base/dautils.cpp:426]: Can a valid name end in '::' ?  If so then I think we would incorrectly thorw here ?

**Comment by mckellyln** [comment]
@jakesmith just one question.  Otherwise looks good.

File: testing/regress/ecl/dynamicoptflag.ecl [hunk 2]
```diff
@@ -34,7 +34,7 @@ count(p);
 output(p(f='NOT'));
 count(p(f='NOT'));
 
-p2 := PRELOAD(dataset(DYNAMIC(prefix + 'no:_such_file_again'), {string10 f}, FLAT, OPT), 2);
+p2 := PRELOAD(dataset(DYNAMIC(prefix + 'no_such_file_again'), {string10 f}, FLAT, OPT), 2);
```
Comment (mckellyln) [inline testing/regress/ecl/dynamicoptflag.ecl:37]: Should we add in a new regression test for this ?

**Review by mckellyln — COMMENTED** [review]
Just a small question

**Comment by richardkchapman** [comment]
What happens if you want to run a local standalone program that refers to (say) 'C:\AUTOEXEC.BAT' ?

File: testing/regress/ecl/dynamicoptflag.ecl (hunk 2 above)
Comment (jakesmith) [reply testing/regress/ecl/dynamicoptflag.ecl:37]: will extend unittests to test invlid paths like this

**Comment by jakesmith** [comment]
>What happens if you want to run a local standalone program that refers to (say) 'C:\AUTOEXEC.BAT' ?

This code is with in the logical file handling helper CDfsLogicalFileName.
I think Roxie with direct physical names, must avoid using that, otherwise it would have already hit errors if the name had e.g. '/' or '\' in them.

To access a physical external file in other engines, it has to be encoded as a logical filename using the external file scope.
There is a helper function ExternalLogicalFileName, which can take a plain external form, and return the external logical name equivalent

File: dali/base/dautils.cpp (hunk 1 above)
Comment (jakesmith) [reply dali/base/dautils.cpp:426]: no, that would already have been rejected as invalid.

**Comment by jakesmith** [comment]
@mckellyln - the regression suite has picked up an issue where there is valid case of single colon, i.e. ~foreign::10.7.8.9:7010::scope1

Need to special case that 'scope'.

**Comment by jakesmith** [comment]
Closing for now. The code that parses lfn's walks over name multiple times, it's a bit silly and needs tidying up.
To correctly spot the form <type> (e.g. foreign) and that it can have different characters could be done with a kludge, but this code needs revisiting to fix properly.

**Pull request closed without merging**